node/db/src/rolling/gc.rs

```diff
@@ -0,0 +1,230 @@
+// Copyright 2019-2023 ChainSafe Systems
+// SPDX-License-Identifier: Apache-2.0, MIT
+
+//!
+//! The current implementation of the garbage collector is a concurrent,
+//! semi-space one.
+//!
+//! ## Design goals
+//! Implement a correct GC algorithm that is simple and efficient for forest
+//! scenarios.
+//!
+//! ## GC algorithm
+//! We choose `semi-space` GC algorithm for simplicity and sufficiency
+//! Besides `semi-space`, `mark-and-sweep` was also considered and evaluated.
+//! However, it's not feasible because of the limitations of the underlying DB
+//! we use, more specifically, limitations in iterating the DB and retrieving the original key. See <https://github.com/paritytech/parity-db/issues/187>
+//!
+//! ## GC workflow
+//! 1. Walk back from the current heaviest tipset to the genesis block, collect
+//! all the blocks that are reachable from the snapshot
+//! 2. writes blocks that are absent from the `current` database to it
+//! 3. delete `old` database(s)
+//! 4. sets `current` database to a newly created one
+//!
+//! ## Correctness
+//! This algorithm considers all blocks that are visited during the snapshot
+//! export task reachable, and ensures they are all transferred and kept in the
+//! current DB space. A snapshot can be used to bootstrap a node from
+//! scratch thus the algorithm is considered appropriate when the post-GC
+//! database contains blocks that are sufficient for exporting a snapshot
+//!
+//! ## Disk usage
+//! During `walk_snapshot`, data from the `old` DB is duplicated in the
+//! `current` DB, which uses extra disk space of up-to 100% of the snapshot file
+//! size
+//!
+//! ## Memory usage
+//! During the data carry-over process, a memory buffer with a fixed capacity is
+//! used to speed up the database write operation
+//!
+//! ## Scheduling
+//! 1. GC is triggered automatically when total DB size is greater than 2x of
+//! the last reachable data size
+//! 2. GC can be triggered manually by `forest-cli db gc` command
+//! 3. There's global GC lock to ensure at most one GC job is running
```

Comment:
```suggestion
//! the last reachable data size
//! 2. GC can be triggered manually by `forest-cli db gc` command
//! 3. There's a global GC lock to ensure at most one GC job is running
```

Reply:
Do we need to specify the chain?

Reply:
Yes for calibnet. But I think it will no longer be needed once the other chain name deduction pr is merged